File: src/saga-desktop/src/net/wombatrpgs/saga/rpg/warheads/EffectMultihit.java
/**
 *  EffectMultihit.java
 *  Created on Apr 26, 2014 3:38:55 PM for project saga-desktop
 *  Author: psy_wombats
 *  Contact: psy_wombats@wombatrpgs.net
 */
package net.wombatrpgs.saga.rpg.warheads;

import net.wombatrpgs.saga.core.SConstants;
import net.wombatrpgs.saga.rpg.Battle;
import net.wombatrpgs.saga.rpg.Chara;
import net.wombatrpgs.saga.rpg.CombatItem;
import net.wombatrpgs.sagaschema.rpg.abil.data.warheads.EffectMultihitMDO;

/**
 * 3-heads, 8-legs, etc
 */
public class EffectMultihit extends EffectCombat {
	
	protected EffectMultihitMDO mdo;
	protected Battle lastUsedBattle;
	protected int sequence[];
	protected int seqIndex;

	/**
	 * Inherited constructor.
	 * @param	mdo				The data to create from
	 * @param	item			The item to create for
	 */
	public EffectMultihit(EffectMultihitMDO mdo, CombatItem item) {
		super(mdo, item);
		this.mdo = mdo;
		String hitsStrings[] = mdo.hits.split(",");
		sequence = new int[hitsStrings.length];
		for (int i = 0; i < hitsStrings.length; i += 1) {
			sequence[i] = Integer.valueOf(hitsStrings[i]);
		}
	}

	/**
	 * @see net.wombatrpgs.saga.rpg.warheads.EffectCombat#calcPower
	 * (Battle, net.wombatrpgs.saga.rpg.Chara)
	 */
	@Override
	protected int calcPower(Battle battle, Chara user) {
		String tab = SConstants.TAB;
		if (lastUsedBattle == battle) {
			seqIndex += 1;
		} else {
			lastUsedBattle = battle;
			seqIndex = 0;
		}
		int hits = sequence[seqIndex];
		battle.println(tab + "Hit " + hits + " times.");
		return statDamage(user, mdo.attackStat, mdo.power) * hits;
	}

	/**
	 * @see net.wombatrpgs.saga.rpg.warheads.EffectCombat#calcDamage
	 * (Battle, int, net.wombatrpgs.saga.rpg.Chara)
	 */
	@Override
	protected int calcDamage(Battle battle, int power, Chara target) {
		int hits = sequence[seqIndex];
		if (mdo.defendStat != null && !weak(target)) {
			power -= (target.get(mdo.defendStat) * hits);
		}
		return power;
	}

	/**
	 * @see net.wombatrpgs.saga.rpg.warheads.EffectCombat#hits
	 * (net.wombatrpgs.saga.rpg.Battle, net.wombatrpgs.saga.rpg.Chara,
	 * net.wombatrpgs.saga.rpg.Chara, float)
	 */
	@Override
	protected boolean hits(Battle battle, Chara user, Chara target, float roll) {
		return true;
	}

}

File: saga-unity/Assets/Scripts/RPG/Effects/EffectMultiHit.cs
using System.Threading.Tasks;

public class EffectMultihit : EffectCombat {

    protected new EffectMultihitData data;

    protected Battle lastUsedBattle;
    protected int[] sequence;
    protected int seqIndex;

    public EffectMultihit(EffectMultihitData data, CombatItem item) : base(data, item) {
        this.data = data;

        var hitsStrings = data.hits.Split(',');
        sequence = new int[hitsStrings.Length];
        for (int i = 0; i < hitsStrings.Length; i += 1) {
            sequence[i] = int.Parse(hitsStrings[i]);
        }
    }

    protected override async Task<int> CalculateAttackPowerAsync(Intent intent) {
        var tab = BattleBox.Tab;
        if (lastUsedBattle == intent.Battle) {
            seqIndex += 1;
        } else {
            lastUsedBattle = intent.Battle;
            seqIndex = 0;
        }
        seqIndex = seqIndex % sequence.Length;
        var hits = sequence[seqIndex];
        await intent.Battle.WriteLineAsync(tab + "Hit " + hits + " times.");
        return CalcStatPower(intent.Actor, data.attackStat, data.power) * hits;
    }

    protected override Task<int> CalculateDamageAsync(Battle battle, int power, Unit target) {
        return Task.FromResult(ReducePowerByDefense(target, data.defendStat, power));
    }

    protected override bool CheckCombatFormulaHit(Battle battle, Unit user, Unit target, float roll) {
        foreach (var effect in battle.GetDefensesForUnit(target)) {
            if (effect.ShieldDodgeBonus > 100) {
                return false;
            }
        }
        return true;
    }

    protected override bool IsPhysical() {
        return data.defendStat == StatTag.DEF;
    }
}